# Importando Bibliotecas

In [457]:
import pandas as pd

# Aquisição de dados

In [458]:
churn_df = pd.read_json('https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json')

churn_df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

# Informações  iniciais do Dataset

In [459]:
churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


# Entendendo as informações presentes no Dataset

* **`customerID`** - ID do cliente
* **`Churn`** - representa se a pessoa deixou a empresa ou não
* **`curtomer`** - objeto com características dos clientes
* **`phone`** - objeto com informações do telefone do cliente
* **`internet`** - objeto com informações da internet do cliente
* **`account`** - objeto com informações da conta do cliente

### Coluna `Churn`

In [460]:
churn_df['Churn'].value_counts()

No     5174
Yes    1869
        224
Name: Churn, dtype: int64

Fica evidente que temos 224 dados da coluna `Churn` que estão vazios. Ou seja, há dados em que a informação chave para a análise desse caso estão não-rotulados. 

In [461]:
no_labeled_churn = churn_df.query("Churn == ''")

no_labeled_churn.head()

customerID Churn                                           customer  \
30   0047-ZHDTW        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
75   0120-YZLQA        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
96   0154-QYHJU        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
98   0162-RZGMZ        {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
175  0274-VVQOQ        {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   

                                               phone  \
30   {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
75    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
96    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
98    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
175  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                              internet  \
30   {'InternetService': 'Fiber optic', 'OnlineSecu...   
75   {'InternetService': 'No', 'OnlineSecurity': 'N...   
96   {'InternetService': 'DSL', 'OnlineSecurity': '...   
98   {'InternetService': 'DSL', 'OnlineSecurity': '...   
175  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                               account  
30   {'Contract': 'Month-to-month', 'PaperlessBilli...  
75   {'Contract': 'Two year', 'PaperlessBilling': '...  
96   {'Contract': 'One year', 'PaperlessBilling': '...  
98   {'Contract': 'Month-to-month', 'PaperlessBilli...  
175  {'Contract': 'One year', 'PaperlessBilling': '...

In [462]:
no_labeled_churn_mask = ~(churn_df['Churn'] == '')

no_labeled_churn_mask

0       True
1       True
2       True
3       True
4       True
        ... 
7262    True
7263    True
7264    True
7265    True
7266    True
Name: Churn, Length: 7267, dtype: bool

## Informações presentes nas colunas que apresentam objetos

### Coluna `customer`
* gender
* SeniorCitizen
* Partner
* Dependents
* tenure

In [463]:
customer_df = pd.json_normalize(data=churn_df['customer'])

customer_df.head()

gender  SeniorCitizen Partner Dependents  tenure
0  Female              0     Yes        Yes       9
1    Male              0      No         No       9
2    Male              0      No         No       4
3    Male              1     Yes         No      13
4  Female              1     Yes         No       3

In [464]:
gender_count = pd.DataFrame(customer_df['gender'].value_counts())

gender_count.rename(columns={'gender': 'Count'}, inplace=True)
gender_count.index.name = 'Gender'

gender_count

Count
Gender       
Male     3675
Female   3592

In [465]:
SeniorCitizen_count = pd.DataFrame(customer_df['SeniorCitizen'].value_counts())

SeniorCitizen_count.rename(columns={'SeniorCitizen': 'Count'}, index={0: 'No', 1: 'Yes'}, inplace=True)
SeniorCitizen_count.index.name = 'SeniorCitizen'

SeniorCitizen_count

Count
SeniorCitizen       
No              6085
Yes             1182

In [466]:
Partner_count = pd.DataFrame(customer_df['Partner'].value_counts())

Partner_count.rename(columns={'Partner': 'Count'}, inplace=True)
Partner_count.index.name = 'Partner'

Partner_count

Count
Partner       
No        3749
Yes       3518

In [467]:
Dependents_count = pd.DataFrame(customer_df['Dependents'].value_counts())

Dependents_count.rename(columns={'Dependents': 'Count'}, inplace=True)
Dependents_count.index.name = 'Dependents'

Dependents_count

Count
Dependents       
No           5086
Yes          2181

In [468]:
tenure_nan_count = pd.DataFrame(customer_df['tenure'].isna().value_counts())

tenure_nan_count.rename(columns={'tenure': 'Count'}, inplace=True)
tenure_nan_count.index.name = 'tenure is NaN'

tenure_nan_count

Count
tenure is NaN       
False           7267

---
### Coluna `phone`
* PhoneService
* MultipleLines

In [469]:
phone_df = pd.json_normalize(data=churn_df['phone'])

phone_df.head()

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No
3          Yes            No
4          Yes            No

In [470]:
PhoneService_count = pd.DataFrame(phone_df['PhoneService'].value_counts())

PhoneService_count.rename(columns={'PhoneService': 'Count'}, inplace=True)
PhoneService_count.index.name = 'PhoneService'

PhoneService_count

Count
PhoneService       
Yes            6560
No              707

In [471]:
MultipleLines_count = pd.DataFrame(phone_df['MultipleLines'].value_counts())

MultipleLines_count.rename(columns={'MultipleLines': 'Count'}, inplace=True)
MultipleLines_count.index.name = 'MultipleLines'

MultipleLines_count

Count
MultipleLines          
No                 3495
Yes                3065
No phone service    707

---
### Coluna `internet`
* InternetService	
* OnlineSecurity	
* OnlineBackup	
* DeviceProtection	
* TechSupport	
* StreamingTV	
* StreamingMovies

In [472]:
internet_df = pd.json_normalize(data=churn_df['internet'])

internet_df.head()

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   
3     Fiber optic             No          Yes              Yes          No   
4     Fiber optic             No           No               No         Yes   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No  
3         Yes             Yes  
4         Yes              No

In [473]:
InternetService_count = pd.DataFrame(internet_df['InternetService'].value_counts())

InternetService_count.rename(columns={'InternetService': 'Count'}, inplace=True)
InternetService_count.index.name = 'InternetService'

InternetService_count

Count
InternetService       
Fiber optic       3198
DSL               2488
No                1581

In [474]:
OnlineSecurity_count = pd.DataFrame(internet_df['OnlineSecurity'].value_counts())

OnlineSecurity_count.rename(columns={'OnlineSecurity': 'Count'}, inplace=True)
OnlineSecurity_count.index.name = 'OnlineSecurity'

OnlineSecurity_count

Count
OnlineSecurity            
No                    3608
Yes                   2078
No internet service   1581

In [475]:
OnlineBackup_count = pd.DataFrame(internet_df['OnlineBackup'].value_counts())

OnlineBackup_count.rename(columns={'OnlineBackup': 'Count'}, inplace=True)
OnlineBackup_count.index.name = 'OnlineBackup'

OnlineBackup_count

Count
OnlineBackup              
No                    3182
Yes                   2504
No internet service   1581

In [476]:
DeviceProtection_count = pd.DataFrame(internet_df['DeviceProtection'].value_counts())

DeviceProtection_count.rename(columns={'DeviceProtection': 'Count'}, inplace=True)
DeviceProtection_count.index.name = 'DeviceProtection'

DeviceProtection_count

Count
DeviceProtection          
No                    3195
Yes                   2491
No internet service   1581

In [477]:
TechSupport_count = pd.DataFrame(internet_df['TechSupport'].value_counts())

TechSupport_count.rename(columns={'TechSupport': 'Count'}, inplace=True)
TechSupport_count.index.name = 'TechSupport'

TechSupport_count

Count
TechSupport               
No                    3582
Yes                   2104
No internet service   1581

In [478]:
StreamingTV_count = pd.DataFrame(internet_df['StreamingTV'].value_counts())

StreamingTV_count.rename(columns={'StreamingTV': 'Count'}, inplace=True)
StreamingTV_count.index.name = 'StreamingTV'

StreamingTV_count

Count
StreamingTV               
No                    2896
Yes                   2790
No internet service   1581

In [479]:
StreamingMovies_count = pd.DataFrame(internet_df['StreamingMovies'].value_counts())

StreamingMovies_count.rename(columns={'StreamingMovies': 'Count'}, inplace=True)
StreamingMovies_count.index.name = 'StreamingMovies'

StreamingMovies_count

Count
StreamingMovies           
No                    2870
Yes                   2816
No internet service   1581

---
### Coluna `account`
* Contract	
* PaperlessBilling	
* PaymentMethod	
* Charges_Monthly	
* Charges_Total

In [480]:
account_df = pd.json_normalize(data=churn_df['account'], sep='_')

account_df.head()

Contract PaperlessBilling     PaymentMethod  Charges_Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   
3  Month-to-month              Yes  Electronic check             98.0   
4  Month-to-month              Yes      Mailed check             83.9   

  Charges_Total  
0         593.3  
1         542.4  
2        280.85  
3       1237.85  
4         267.4

In [481]:
account_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Contract          7267 non-null   object 
 1   PaperlessBilling  7267 non-null   object 
 2   PaymentMethod     7267 non-null   object 
 3   Charges_Monthly   7267 non-null   float64
 4   Charges_Total     7267 non-null   object 
dtypes: float64(1), object(4)
memory usage: 284.0+ KB


In [482]:
(account_df['Charges_Total'] == ' ').value_counts()

False    7256
True       11
Name: Charges_Total, dtype: int64

Logo, é possível perceber que, além de estar como tipo objeto, nossa coluna `Charges_Total` possui valores vazios.

In [483]:
Contract_count = pd.DataFrame(account_df['Contract'].value_counts())

Contract_count.rename(columns={'Contract': 'Count'}, inplace=True)
Contract_count.index.name = 'Contract'

Contract_count

Count
Contract             
Month-to-month   4005
Two year         1743
One year         1519

In [484]:
PaperlessBilling_count = pd.DataFrame(account_df['PaperlessBilling'].value_counts())

PaperlessBilling_count.rename(columns={'PaperlessBilling': 'Count'}, inplace=True)
PaperlessBilling_count.index.name = 'PaperlessBilling'

PaperlessBilling_count

Count
PaperlessBilling       
Yes                4311
No                 2956

In [485]:
PaymentMethod_count = pd.DataFrame(account_df['PaymentMethod'].value_counts())

PaymentMethod_count.rename(columns={'PaymentMethod': 'Count'}, inplace=True)
PaymentMethod_count.index.name = 'PaymentMethod'

PaymentMethod_count

Count
PaymentMethod                   
Electronic check            2445
Mailed check                1665
Bank transfer (automatic)   1589
Credit card (automatic)     1568

In [486]:
Charges_Monthly_nan_count = pd.DataFrame(account_df['Charges_Monthly'].isna().value_counts())

Charges_Monthly_nan_count.rename(columns={'Charges_Monthly': 'Count'}, inplace=True)
Charges_Monthly_nan_count.index.name = 'Charges_Monthly'

Charges_Monthly_nan_count

Count
Charges_Monthly       
False             7267

In [487]:
Charges_Total_nan_count = pd.DataFrame(account_df['Charges_Total'].isna().value_counts())

Charges_Total_nan_count.rename(columns={'Charges_Total': 'Count'}, inplace=True)
Charges_Total_nan_count.index.name = 'Charges_Total'

Charges_Total_nan_count

Count
Charges_Total       
False           7267

# Juntando todas as informações em um só dataset


In [488]:
churn_df = pd.concat([churn_df.drop(columns=['customerID', 'customer', 'phone', 'internet', 'account']), 
                      customer_df, phone_df, internet_df, account_df], axis='columns')

churn_df.head()

Churn  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0    No  Female              0     Yes        Yes       9          Yes   
1    No    Male              0      No         No       9          Yes   
2   Yes    Male              0      No         No       4          Yes   
3   Yes    Male              1     Yes         No      13          Yes   
4   Yes  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   
3            No     Fiber optic             No          Yes              Yes   
4            No     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      PaymentMethod  Charges_Monthly Charges_Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85  
3  Electronic check             98.0       1237.85  
4      Mailed check             83.9         267.4

In [489]:
churn_df = churn_df[no_labeled_churn_mask]

churn_df.head()

Churn  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0    No  Female              0     Yes        Yes       9          Yes   
1    No    Male              0      No         No       9          Yes   
2   Yes    Male              0      No         No       4          Yes   
3   Yes    Male              1     Yes         No      13          Yes   
4   Yes  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   
3            No     Fiber optic             No          Yes              Yes   
4            No     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      PaymentMethod  Charges_Monthly Charges_Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85  
3  Electronic check             98.0       1237.85  
4      Mailed check             83.9         267.4

In [490]:
no_value_Charges_Total_mask = ~(churn_df['Charges_Total'] == ' ')

no_value_Charges_Total_mask

0       True
1       True
2       True
3       True
4       True
        ... 
7262    True
7263    True
7264    True
7265    True
7266    True
Name: Charges_Total, Length: 7043, dtype: bool

In [491]:
churn_df = churn_df[no_value_Charges_Total_mask]

churn_df.head()

Churn  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0    No  Female              0     Yes        Yes       9          Yes   
1    No    Male              0      No         No       9          Yes   
2   Yes    Male              0      No         No       4          Yes   
3   Yes    Male              1     Yes         No      13          Yes   
4   Yes  Female              1     Yes         No       3          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   
3            No     Fiber optic             No          Yes              Yes   
4            No     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      PaymentMethod  Charges_Monthly Charges_Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85  
3  Electronic check             98.0       1237.85  
4      Mailed check             83.9         267.4

In [492]:
churn_df['Charges_Total'].iloc[0, :] = churn_df['Charges_Total'].apply(float)

In [493]:
churn_df.reset_index(inplace=True)
churn_df.drop(columns=['index'], inplace=True)

In [494]:
churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Churn             7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 


Finalizando assim, o tratamento inicial da nossa base de dados